src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSWebWorker.cs

```diff
@@ -30,220 +30,205 @@ public static Task RunAsync(Func<Task> body)
                 }
                 catch (Exception ex)
                 {
-                    SendWhenException(parentContext, tcs, ex);
+                    PropagateCompletionAndDispose(Task.FromException<T>(ex));
                 }
-
-            });
-            JSHostImplementation.SetHasExternalEventLoop(t);
-            t.Start();
-            return tcs.Task;
-        }
-
-        #region posting result to the original thread when handling exception
-
-        private static void PostWhenCancellation(SynchronizationContext ctx, TaskCompletionSource tcs)
-        {
-            try
-            {
-                ctx.Send((_) => tcs.SetCanceled(), null);
-            }
-            catch (Exception e)
-            {
-                Environment.FailFast("JSWebWorker.RunAsync failed", e);
             }
-        }
 
-        private static void PostWhenCancellation<T>(SynchronizationContext ctx, TaskCompletionSource<T> tcs)
-        {
-            try
-            {
-                ctx.Send((_) => tcs.SetCanceled(), null);
-            }
-            catch (Exception e)
+            // run actions on correct thread
+            private void PropagateCompletionAndDispose(Task result)
             {
-                Environment.FailFast("JSWebWorker.RunAsync failed", e);
-            }
-        }
+                _result = result;
 
-        private static void SendWhenDone(SynchronizationContext ctx, TaskCompletionSource tcs, Task done)
-        {
-            try
-            {
-                ctx.Send((_) =>
+                _cancellationRegistration?.Dispose();
+                _cancellationRegistration = null;
+
+                if (Thread.CurrentThread == _thread)
                 {
-                    PropagateCompletion(tcs, done);
-                }, null);
-            }
-            catch (Exception e)
-            {
-                Environment.FailFast("JSWebWorker.RunAsync failed", e);
-            }
-        }
+                    // if cancelation was requested, the JSSynchronizationContext will stop the thread
+                    if (_jsSynchronizationContext != null && !_cancellationToken.IsCancellationRequested)
+                    {
+                        // this will lead to __emscripten_thread_exit() later
+                        _jsSynchronizationContext.UninstallWebWorkerInterop();
+                    }
+                    _jsSynchronizationContext = null;
 
-        private static void SendWhenException(SynchronizationContext ctx, TaskCompletionSource tcs, Exception ex)
-        {
-            try
-            {
-                ctx.Send((_) => tcs.SetException(ex), null);
-            }
-            catch (Exception e)
-            {
-                Environment.FailFast("JSWebWorker.RunAsync failed", e);
-            }
-        }
+                    // propagate the result on thread pool, rather than the JSWebWorker
```

Comment:
should we push it to the main thread queue instead? i'm ok with this but it feels like the main thread may be the correct place

Reply:
I'm still considering if the UI thread would be non-managed. Also I'm considering that Blazor UI events should be pushed into thread-pool asap. Now that we are running the unit tests in the TP, the UI thread is essentially idle and that solved so many problems.